### PR TITLE
Add support to nested controller

### PR DESCRIPTION
**Pull request by maximenoel8**
## What does this PR change?

Add bastion support to run ssh command or push files.

### CODE REVIEW

**Review by juliogonzalez — COMMENTED** [review]
> Currently, I'm using terracucumber-cli to deploy SUMA 4.2 with MU on server without using controller. cucumber_put is not compatible with AWS yet.

Strictly speaking this is not about being compatible with AWS, but about being compatible with bastion hosts. It would fail outside AWS as well, unless I am missing something.

My question here is: why do we need this change? You could change the pipeline not to pass the parameter for `custom-repositories`, until we fix terracumber to be compatible with bastion hosts, right? That should avoid uploading custom repositories.

**Comment by maximenoel8** [comment]
I already changed my pipeline to be compatible with terracumber so I would like to keep it like that.

I added the bastion support for AWS. Tell me what you think ? @juliogonzalez

File: terracumber-cli [hunk 1]
```diff
@@ -79,6 +80,10 @@ def parse_args():
                                                       libvirt is used by default""",
                         choices=['libvirt', 'aws', 'null'], default="libvirt",
                         dest='sumaform_backend')
+    parser.add_argument('--bastion_key', help="""bastion key to be use""",
+                        default="/home/jenkins/.ssh/testing-suma.pem",
```
Comment (juliogonzalez) [inline terracumber-cli]:
```suggestion
                        default="~/.ssh/testing-suma.pem",
```
If I am not wrong, you will need ti expand `~`. I suggest doing it at the `parse_args` function, before returning.

File: terracumber-cli [hunk 2]
```diff
@@ -79,6 +80,10 @@ def parse_args():
                                                       libvirt is used by default""",
                         choices=['libvirt', 'aws', 'null'], default="libvirt",
                         dest='sumaform_backend')
+    parser.add_argument('--bastion_key', help="""bastion key to be use""",
```
Comment (juliogonzalez) [inline terracumber-cli]:
```suggestion
    parser.add_argument('--bastion_ssh_key', help="""bastion key to be use""",
```

File: terracumber-cli [hunk 3]
```diff
@@ -314,6 +345,10 @@ def main():
         print("Running terraform...")
         results['terraform'] = run_terraform(args, tf_vars)
 
+    if args.sumaform_backend == 'aws' and (args.runstep == 'cucumber' or args.custom_repositories):
```
Comment (juliogonzalez) [inline terracumber-cli]: Let's not make the bastion related to the backend being used. It should work as well if we add Azure tomorrow. And it should work if for whatever reason we don't use a bastion.

I'd suggest you check if `get_bastion_hostname` is returning `None` when the bastion does not exist, and if so you check here if it's needed or not.

You will also need to run when `args.runall` is true.

File: terracumber-cli [hunk 4]
```diff
@@ -108,6 +113,24 @@ def read_config(tf_file):
     return config
 
 
+def open_bastion_channel(args, tf_vars, bst_creds, ctl_creds):
+    """Update bastion and controller update"""
+    ctl = get_controller_hostname(args, tf_vars)
+    ctl_creds['hostname'] = ctl
```
Comment (juliogonzalez) [inline terracumber-cli]: As car as I can tell, you are not using `ctl` later, so you should be able to do:
```suggestion
    ctl_creds['hostname'] = get_controller_hostname(args, tf_vars)
```

File: terracumber-cli [hunk 5]
```diff
@@ -108,6 +113,24 @@ def read_config(tf_file):
...
+    bst = get_bastion_hostname(args, tf_vars)
+    bst_creds['hostname'] = bst
```
Comment (juliogonzalez) [inline terracumber-cli]: Same as before

File: terracumber-cli [hunk 6]
```diff
@@ -108,6 +113,24 @@ def read_config(tf_file):
...
+    """Create connection to bastion"""
+    bastion = paramiko.SSHClient()
+    bastion.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    bastion.connect(**bst_creds)
+    bastion_transport = bastion.get_transport()
+    """Creation channel between bastion and controller"""
+    dest_addr = (ctl_creds["hostname"], 22)  # edited#
```
Comment (juliogonzalez) [inline terracumber-cli]: What does the comment mean?

File: terracumber-cli [hunk 7]
```diff
@@ -108,6 +113,24 @@ def read_config(tf_file):
...
+    local_addr = (bst_creds["hostname"], 22)  # edited#
```
Comment (juliogonzalez) [inline terracumber-cli]: Same.

**Review by juliogonzalez — CHANGES_REQUESTED** [review]
Looks good in general. I left some comments to make it more generic, as well as some questions.

**Comment by maximenoel8** [comment]
Thanks for the review.
I updated with comments. Tell me if it's better @juliogonzalez.

File: terracumber-cli [hunk 8]
```diff
@@ -79,6 +80,10 @@ def parse_args():
                                                       libvirt is used by default""",
                         choices=['libvirt', 'aws', 'null'], default="libvirt",
                         dest='sumaform_backend')
+    parser.add_argument('--bastion_ssh_key', help="""bastion key to be use""",
+                        default="~/.ssh/testing-suma.pem",
```
Comment (juliogonzalez) [inline terracumber-cli]: Did you check  if `~` is understood by `bastion.connect`? Not sure if you need to expand it or not.

File: terracumber-cli [hunk 9]
```diff
@@ -314,6 +346,11 @@ def main():
         print("Running terraform...")
         results['terraform'] = run_terraform(args, tf_vars)
 
+    if get_bastion_hostname(args, tf_vars) is not None and (
```
Comment (juliogonzalez) [inline terracumber-cli]: You can make this shorter. If bastion is not None, then it's true:

```suggestion
    if get_bastion_hostname(args, tf_vars) and (
```

File: terracumber-cli [hunk 10]
```diff
@@ -108,6 +113,21 @@ def read_config(tf_file):
...
+    """Set bastion and controller hostname"""
+    ctl_creds['hostname'] = get_controller_hostname(args, tf_vars)
+    bst_creds['hostname'] = get_bastion_hostname(args, tf_vars)
+    """Open connection with bastion"""
+    bastion = paramiko.SSHClient()
+    bastion.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    bastion.connect(**bst_creds)
+    bastion_transport = bastion.get_transport()
+    """Create port forward between bastion and controller"""
+    bastion_channel = bastion_transport.open_channel("direct-tcpip", (ctl_creds["hostname"], 22),
```
Comment (juliogonzalez) [inline terracumber-cli]: At some point later we should probably make the port for the controller, and the port for the bastion configurable, but it's OK for now.

File: terracumber-cli [hunk 11]
```diff
@@ -295,11 +326,12 @@ def main():
     # Pending: Make credentials and port configurable, and allow keypairs usage both from file
     # and agent (already supported by terraform.cucumber)
     ctl_creds = {'hostname': None, 'username': 'root',
-                 'port': 22, 'password': 'linux'}
+                 'port': 22, 'password': 'linux', 'sock': None}
```
Comment (juliogonzalez) [inline terracumber-cli:329]: Just curious, why do we need this?

**Review by juliogonzalez — CHANGES_REQUESTED** [review]
Looks good! Just one minor change, and I'd say we are ready to go.

What does it mean "Add support to nested cucumber"? This is about adding support for a bastion host, right?

File: terracumber-cli (hunk 11 above)
Comment (maximenoel8) [reply terracumber-cli:329]: It's the variable I use then to set my bastion port forward. You can see at lane 352 that I change it for the channel  created before.

File: terracumber-cli (hunk 8 above)
Comment (maximenoel8) [reply terracumber-cli]: No it was failing. I set the full path in my pipeline. I can completely remove the default value to have no mention of jenkins env.

**Comment by maximenoel8** [comment]
> Looks good! Just one minor change, and I'd say we are ready to go.
> 
> What does it mean "Add support to nested cucumber"? This is about adding support for a bastion host, right?

On AWS, controller is deployed on private network so it's not possible to reach it directly. So I call it nested but yes it's to support the bastion.

File: terracumber-cli [hunk 12]
```diff
@@ -79,6 +80,9 @@ def parse_args():
                                                       libvirt is used by default""",
                         choices=['libvirt', 'aws', 'null'], default="libvirt",
                         dest='sumaform_backend')
+    parser.add_argument('--bastion_ssh_key', help="""bastion key to be use""",
+                        dest='bastion_ssh_key'
+                        )
```
Comment (juliogonzalez) [inline terracumber-cli]: If you removed the default value, you can move this to line 84. I don't remember how `argparse` behaves, but I guess that this `args.bastion_ssh_key` will be `None` if the key is not specified, so lines 332/333 won't fail, right?

**Comment by maximenoel8** [comment]
@juliogonzalez Tell me if the last changes are enough and if I can merge the PR. 

Thanks for your help.

**Review by juliogonzalez — APPROVED** [review]
Argh, I am sorry. I wrote this yesterday, but didn't submit the review. If you can fix this, we are ready to merge (also please change the PR topic, this is not about nested controllers, but about using a bastion)

**Comment by maximenoel8** [comment]
@juliogonzalez I added a error raise if bastion ssh key is missing. I can not merge this PR could you give me the rights or do it yourself please ?

**Review by juliogonzalez — APPROVED** [review]
Looks good, but let's not merge today, just in case we break all the other testsuites for the whole weekend.

I am going to give you rights, and you can **coordinate** with the RRTG and merge on Monday, so if this breaks testsuites, you can quickly rollback.

**Pull request merged**